spaceapi/src/utils.rs

```diff
@@ -88,6 +108,30 @@ impl<T> Optional<T> {
         }
     }
 
+    /// Returns `Absent` if the optional is `Absent`, otherwise calls `f` with the
+    /// wrapped value and returns the result.
+    ///
+    /// Some languages call this operation flatmap.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use spaceapi::utils::Optional;
+    /// # use spaceapi::utils::Optional::{Value,Absent};
+    /// fn sq(x: u32) -> Optional<u32> { Value(x * x) }
+    /// fn nope(_: u32) -> Optional<u32> { Absent }
+    ///
+    /// assert_eq!(Value(2).and_then(sq).and_then(sq), Value(16));
+    /// assert_eq!(Value(2).and_then(sq).and_then(nope), Absent);
+    /// assert_eq!(Value(2).and_then(nope).and_then(sq), Absent);
+    /// assert_eq!(Absent.and_then(sq).and_then(sq), Absent);
+    /// ```
+    pub fn and_then<U, F: FnOnce(T) -> Optional<U>>(self, f: F) -> Optional<U> {
```

Comment:
The difference is in the closure that gets passed.
 * In `map` the closure returns `T` and `map` packs it into an `Optional<T>`: `F: FnOnce(T) -> U`
 * In `and_then` the closure already returns `Optional<T>` and `and_then` just returns that: ` FnOnce(T) -> Optional<U>`

At first I used `map` in https://github.com/coredump-ch/spaceapi/pull/54/files#diff-371a269f64ba2af763532985ef5c8bf8R95 (That's why I implemented it) and got an `Optional<Optional<U>>` back. So I changed to `and_then`.